File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/xmi/PrettyPrinter.java
package org.kevoree.modeling.api.xmi;

import org.kevoree.modeling.api.Callback;
import org.kevoree.modeling.api.util.Helper;

public class PrettyPrinter implements Callback<Throwable> {
        private SerializationContext context;

        public PrettyPrinter(SerializationContext context) {
            this.context = context;
        }

        @Override
        public void on(Throwable throwable) {
            if (throwable != null) {
                context.finishCallback.on(null,throwable);
            } else {
                context.printer.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
                context.printer.append("<" + XMIModelSerializer.formatMetaClassName(context.model.metaClass().metaName()).replace(".", "_"));
                context.printer.append(" xmlns:xsi=\"http://wwww.w3.org/2001/XMLSchema-instance\"");
                context.printer.append(" xmi:version=\"2.0\"");
                context.printer.append(" xmlns:xmi=\"http://www.omg.org/XMI\"");

                int index = 0;
                while (index < context.packageList.size()) {
                    context.printer.append(" xmlns:" + context.packageList.get(index).replace(".", "_") + "=\"http://" + context.packageList.get(index) + "\"");
                    index++;
                }
                context.model.visitAttributes(context.attributesVisitor);
                Helper.forall(context.model.metaReferences(), new NonContainedReferencesCallbackChain(context, context.model), new Callback<Throwable>() {
                    @Override
                    public void on(Throwable err) {
                        if (err == null) {
                            context.printer.append(">\n");
                            Helper.forall(context.model.metaReferences(), new ContainedReferencesCallbackChain(context, context.model), new Callback<Throwable>() {
                                @Override
                                public void on(Throwable containedRefsEnd) {
                                    if (containedRefsEnd == null) {
                                        context.printer.append("</" + XMIModelSerializer.formatMetaClassName(context.model.metaClass().metaName()).replace(".", "_") + ">\n");
                                        context.finishCallback.on(context.printer.toString(),null);
                                    } else {
                                        context.finishCallback.on(null,containedRefsEnd);
                                    }
                                }
                            });
                        } else {
                            context.finishCallback.on(null,err);
                        }
                    }
                });
            }
        }
    }

File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/xmi/PrettyPrinter.ts
///<reference path="../Callback.ts"/>
///<reference path="../util/Helper.ts"/>

class PrettyPrinter implements Callback<Throwable> {

  private context: SerializationContext = null;

  constructor(context: SerializationContext) {
    this.context = context;
  }

  public on(throwable: Throwable): void {
    if (throwable != null) {
      this.context.finishCallback.on(null, throwable);
    } else {
      this.context.printer.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
      this.context.printer.append("<" + XMIModelSerializer.formatMetaClassName(this.context.model.metaClass().metaName()).replace(".", "_"));
      this.context.printer.append(" xmlns:xsi=\"http://wwww.w3.org/2001/XMLSchema-instance\"");
      this.context.printer.append(" xmi:version=\"2.0\"");
      this.context.printer.append(" xmlns:xmi=\"http://www.omg.org/XMI\"");
      var index: number = 0;
      while (index < this.context.packageList.size()){
        this.context.printer.append(" xmlns:" + this.context.packageList.get(index).replace(".", "_") + "=\"http://" + this.context.packageList.get(index) + "\"");
        index++;
      }
      this.context.model.visitAttributes(this.context.attributesVisitor);
      Helper.forall(this.context.model.metaReferences(), new NonContainedReferencesCallbackChain(this.context, this.context.model), 
        public on(err: Throwable): void {
          if (err == null) {
            this.context.printer.append(">\n");
            Helper.forall(this.context.model.metaReferences(), new ContainedReferencesCallbackChain(this.context, this.context.model), 
              public on(containedRefsEnd: Throwable): void {
                if (containedRefsEnd == null) {
                  this.context.printer.append("</" + XMIModelSerializer.formatMetaClassName(this.context.model.metaClass().metaName()).replace(".", "_") + ">\n");
                  this.context.finishCallback.on(this.context.printer.toString(), null);
                } else {
                  this.context.finishCallback.on(null, containedRefsEnd);
                }
              }

);
          } else {
            this.context.finishCallback.on(null, err);
          }
        }

);
    }
  }

}

